### PR TITLE
normalizing for 3d model

### PR DESCRIPTION
**Pull request by Shackmeister**

### CODE REVIEW

**Comment by poeschlr** [comment]
Could you also add the prefix ${KISYS3DMOD}/ to the 3D shapes path?

**Comment by poeschlr** [comment]
If you find time, it would be nice if you could make this footprint KLC compliant. (See travis log for details.)

**Comment by Shackmeister** [comment]
and thats all the fixing Im gonna do for now

**Comment by poeschlr** [comment]
Wow that's a lot of changes. I will have a look at that within the next few hours.

**Comment by Shackmeister** [comment]
Yeah went with as many KLC fixes as I could without redrawing too much
switches_dip

**Comment by poeschlr** [comment]
Could you confirm that i interpreted your changes correctly?
Mostly you added ${KISYS3DMOD}/ to the 3D shapes path and the second reference on fab layer.
But for some footprints you did more:
The main thing i noticed is that for parts where you added the courtyard, the courtyard is not done correctly. It should be 0.25mm outside of the part body and pads. (whatever is larger)

- KSA_Tactile_SPST
  - moved it such that pin 1 is on origin
  - added courtyard
  - might it be a good idea to rotate it by 10 degree around pin 1 such that pin one is on the top left?
- SW_Matias_Quiet
  - removed ',' from tags
  - added courtyard
    - **The courtyard should be a bit bigger. It should be larger than the body of the component.**
- SW_Micro_SPST
  - Added courtyard
    - **Assuming the silk screen represents the size of the body outline, the courtyard should be outside.**
- SW_Micro_SPST_Angled
  - Added courtyard
    - **The same as with other components, the courtyard needs to be outside of the complete body not only outside of the pads.**
- SW_NKK_G1xJP
  - Simplified silk screen
  - moved such that pin 1 is at origin
  - added courtyard.
    - again: **courtyard clearance is not the same all around the part. on the sides it seems to be 0**
- SW_PUSH-12mm
  - moved such that pin 1 is at origin
  - added courtyard.
    - again: **courtyard on top and bottom (y axis) is not correct. Should be outside of the part body.**
- SW_PUSH
  - moved such that pin 1 is at origin
  - added courtyard.
    - again: **courtyard on top and bottom (y axis) is not correct. Should be outside of the part body.**
  - added 3d model shape offset (At was 0,0,0 before your change. Now it is 0.2,-0.1,0)
    - We should probably set this back to 0,0,0. The model itself should be moved. (I assume this is because you moved the complete model)
- SW_PUSH_SMALL
  - moved such that pin 1 is at origin 
    - maybe it would be better to move them such that the left pin 1 is on origin.
  - added courtyard.
    - again: **courtyard on top and bottom (y axis) is not correct. Should be outside of the part body.**
- SW_PUSH_x2
  - moved courtyard
    - but again: **courtyard needs to be the same all around the part. (0.25mm outside of the body)**
- SW_SPDT
  - moved such that pin 1 is on the origin
  - not sure if you touched the courtyard here.
    - It is definitely too far away. (it looks like 0.5mm)
  - It might be a good idea to remove all silk features that are above pads.

**Comment by Shackmeister** [comment]
I believe this is correct, I used the --fix option in the check script and didnt notice all the changes, Ill fix them

> might it be a good idea to rotate it by 10 degree around pin 1 such that pin one is on the top left?
you mean -90 degrees?

**Comment by Shackmeister** [comment]
@poeschlr I made some more changes to KSA_Tactile_SPST can you please check if I got it right?

**Comment by poeschlr** [comment]
Looks better. The courtyard on the "bottom" should be at y=8.759 (Copper of the pad is "lower" than the fab line you seem to have used as reference.)
Also is this buttons body really moved towards the "bottom" instead of centered compared to the center of the pins?

(With bottom here i mean the bottom of the screen = towards positive y)

**Comment by Shackmeister** [comment]
@poeschlr did you check after the last commit?

**Comment by poeschlr** [comment]
Apparently not. Looks good now.

**Comment by Shackmeister** [comment]
any notes before I move on to the rest? :)

**Comment by poeschlr** [comment]
The rectangle on the fab layer has a line thickness of 0.15. All other footprints have 0.1mm line width.
Other than that it looks good.

**Comment by Shackmeister** [comment]
lacking some info on the SW_Micro_SPST and SW_Micro_SPST_Angled

**Comment by poeschlr** [comment]
It seems to be similar to a pin header with a jumper?
(At least from the pin pitch and space requirements. I would also read the silk screen this way.)
I think some time ago i have also seen switches that could be fit in such a space. But i could not find one now. 

That's the problem with generic footprints. They might have made sense for the person who designed this footprint but without ordering information nobody else will find out what they are.
(Maybe we could ask over at the forum if someone knows about a component that fits this part.)

Edit: found something similar http://www.farnell.com/datasheets/58539.pdf?_ga=2.66749482.528742633.1495557766-662385333.1495547195

**Comment by Shackmeister** [comment]
Hmm the SPST is throwing me off.. Why 3 pins?

**Comment by poeschlr** [comment]
Maybe it is really reserved to have a place where to put the jumper while it is not connected. (I have no other idea of why it might be needed otherwise.)

**Comment by Shackmeister** [comment]
@CarlPoirier Hey Carl it seems you were the one to add these 2 footprints, do you happen to remember anything else about them?

**Comment by Shackmeister** [comment]
the 3d models
switches

**Comment by Shackmeister** [comment]
SW_PUSH and SW_PUSH_X2 is also lacking information

**Comment by poeschlr** [comment]
I'm not really surprised to be honest. This lib seems to be a bit of a mess. Edit: sorry for the closing/reopening. (clicked the wrong button there.)

**Comment by Shackmeister** [comment]
Would it make sense to just remove the ones without sufficient info?

**Comment by poeschlr** [comment]
I would be for removing files that have no information of what they are intended to be. Electronics design veterans will never use them and novices might fall into the trap. 
This means no one will complain that such footprints are suddenly missing but it might prevent a new user form using a footprint where they are never able to find a correct component that fits this footprint. @SchrodingersGat and @jkriege2 what do you guys think about this?

**Comment by jkriege2** [comment]
For me that would be OK ... which are we talking about?

JAN

**Comment by SchrodingersGat** [comment]
That sounds appropriate to me.

**Comment by CarlPoirier** [comment]
I did not draw these and I have no idea where they come from. I vote for removal.

**Comment by Shackmeister** [comment]
footprints to remove:
SW_Micro_SPST
SW_Micro_SPST_Angled
SW_PUSH
SW_PUSH_X2
SW_SPDT

SW_PUSH_SMALL should be renamed

anybody disagreeing before I remove them?

**Comment by Shackmeister** [comment]
@poeschlr @jkriege2 @SchrodingersGat you agree on the deletions?

also should the SW_Tactile_SPST_Angled have some different variations depending on the length of the button? (and seperate 3d models)

**Comment by SchrodingersGat** [comment]
@shackmeister I agree with removing these.

**Comment by Shackmeister** [comment]
Will hopefully fix the last ones tomorrow So it can be merges

**Comment by Shackmeister** [comment]
@poeschlr This should be Mergeable now
I still have a few other footprints I would like to add but that should be in another PR :)
thanks for the help

**Comment by poeschlr** [comment]
There are still some problems.

Push_E-Switch_KS01Q01
  - 3d shape offset

All sw_dip footprints have a rotation of 90 around the z axis.

SW_DIP_*_Piano
  - Pin 1 marker is currently only on fab layer

SW_DIP_*_Slide_LowProfile
  - Silk body outline is below the part -> pin 1 marker not visible after assembly

SW_Matias_Quiet
  - This footprint has two courtyard rectangles. remove the inner one

SW_NKK_G1xJP
  - 3d shape scaling and position

SW_PUSH-12mm
  - 3d shape scaling and position

SW_PUSH_6mm*
  - 3d shape scaling and position

SW_PUSH_SMALL
  - you wanted to rename it didn't you?

SW_Tactile_SKHH_Angled
  - silk screen lines not touching on the corners.

SW_Tactile_SPST_Angled*
  - silk screen is quite close to pads. (touching copper -> within the masked area for all fabs.)

**Comment by Shackmeister** [comment]
Push_E-Switch_KS01Q01
- 3D offset is made for maintaining compatability with the current 3d model

SW_DIP_*_Piano
- fixed

SW_DIP_*_Slide_LowProfile
- fixed

SW_Matias_Quiet
- fixed

SW_NKK_G1xJP
- fixed position to work with the model, scaling is like this for the model to work

SW_PUSH-12mm
- 3D offset and scaling is made for maintaining compatability with the current 3d model

SW_PUSH_6mm*
- 3D offset and scaling is made for maintaining compatability with the current 3d model (but I might add some new model in the near future)

SW_PUSH_SMALL
- fixed

SW_Tactile_SKHH_Angled
- fixed

SW_Tactile_SPST_Angled*
- fixed

**Comment by poeschlr** [comment]
> How would you suggest to make the pin one marker? (a screenshot would be perfect)

You could mark the corner like we did for some connectors or what is done for some dc/dc converters. You could also just put a small circle there.
screenshot from 2017-05-31 23-18-24

**Comment by Shackmeister** [comment]
all issues above should be fixed now (except 3d offset scaling etc which I believe shouldnt be fixed before a new model is provided)
switches_dip

**Comment by poeschlr** [comment]
In SW_Tactile_SKHH_Angled you have a very large distance between the pads and the lines on silk.
In SW_Tactile_SPST_Angled* this distance is very small.
I would make them similar to this:
screenshot from 2017-06-01 09-23-26
The footprint i created for that (renamed to .txt because .kicad_mod is not allowed in a github comment.)
[SW_Tactile_SPST_Angled_PTS645Vx31-2LFS.txt](https://github.com/KiCad/Buttons_Switches_THT.pretty/files/1043957/SW_Tactile_SPST_Angled_PTS645Vx31-2LFS.txt)

While creating my footprint as an example i noticed that some of your silk screen lines are not straight. The long line which i segmented into 3 lines had its left y = 0.97028 and the right y=0.97. Same with x coordinate. (The line that goes downwards from this corner is not straight as well.) 
In short: This left corner is not on a 0.01 grid but all other points are on this grid.

**Comment by Shackmeister** [comment]
Ill have a look at SW_Tactile_SKHH_Angled today
I aligned SW_Tactile_SPST_Angled_* to your footprints and checked everything is on grid

**Comment by Shackmeister** [comment]
@poeschlr How did you get the pads to show like that in the footprint editor?

**Comment by poeschlr** [comment]
In open gl mode there is a button in the left toolbar that switches between outline and fill mode for pads.
(This button also exists in pcb_new. pcb_new has similar buttons for vias, tracks and copper zones.)
screenshot from 2017-06-01 20-54-04

**Comment by Shackmeister** [comment]
@poeschlr SKHH has now been overhauled, quite alot of the measurements in the datasheet didnt comply with the footprint

**Comment by poeschlr** [comment]
The SKHH looks quite good. (I personally might use a bigger silkscreen clearance but i would accept it as is. Here my suggested footprint: [SW_Tactile_SKHH_Angled.kicad_mod.txt](https://github.com/KiCad/Buttons_Switches_THT.pretty/files/1046090/SW_Tactile_SKHH_Angled.kicad_mod.txt) Remove the txt ending if you want to view it.)

But the SW_Tactile_Straight* footprints have different clearances for different silk screen features.
If you could "normalize" this, i think this pull request is ready to be merged.
screenshot from 2017-06-01 23-11-43

**Comment by Shackmeister** [comment]
@poeschlr my editor doesnt show the most outer ring of the pad :s

**Comment by poeschlr** [comment]
Are you in legacy? The outer ring is the solder mask. In the footprint editor the default clearance is shown. (I think 0.2mm) Sadly the mask layer is only shown in open gl.

**Comment by Shackmeister** [comment]
nope in openGL 4.06
skaermbillede fra 2017-06-01 23-46-56

**Comment by Shackmeister** [comment]
got it
the soldermask clearance was set to 0

**Comment by poeschlr** [comment]
Now i have a question: Where can one set the soldermask clearance such that it changes the display in the footprint editor? (Might be useful for checking pull requests. I would like to set it to 0.25mm and visually check the courtyard clearance.)

**Comment by Shackmeister** [comment]
fixed the issues

I did -> Footprint Properties -> (right lower corner) solder mask clearance

**Comment by poeschlr** [comment]
Ok i assume you use a nightly build version?
I think to remember that this is ignored in the stable version.

Also make sure that this value is set to 0 for submitting footprints. This way the project setting is used for the solder mask clearance. (As far as i can tell it is set to 0 in the footprints commited to this pull request.)

Now the important part: Thanks for your patience with my pedantic requests and for this wonderful contribution.

**Comment by Shackmeister** [comment]
FINALLY! 
nope 4.06 stable :)
and I made sure to set it to zero
glad to help :) I like that you guys set the bar high for the libraries, especially when you answer quite fast and are patient and helpful :)

can you please check the Radial caps PR while you're here? :D

**Comment by poeschlr** [comment]
I will look at that pull request next. (Not sure if i get finished today. But at the latest i will have a result tomorrow morning.)

I have 4.0.2 on this machine and a nightly at the office. (I use this machine for pull request reviews and the office machine for my projects.)
